cmd/gvisor-website/main.go

```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
 	})))
 }
 
+// registerRepo registers the repository handler.
+func registerRepo(mux *http.ServeMux) {
+	if mux == nil {
+		mux = http.DefaultServeMux
+	}
+
+	mux.Handle("/gvisor/git-upload-pack", hostRedirectHandler(http.HandlerFunc(gitUploadPack)))
+	mux.Handle("/gvisor/info/refs", hostRedirectHandler(http.HandlerFunc(infoRefs)))
+}
+
+func gitUploadPack(w http.ResponseWriter, r *http.Request) {
+	target, err := url.Parse(upstreamGitUploadPack)
+	if err != nil {
+		panic(err) // Should never happen.
+	}
+	// Proxy to the upstream repository.
+	proxy := &httputil.ReverseProxy{
+		Director: func(req *http.Request) {
+			req.URL.Scheme = target.Scheme
+			req.URL.Host = target.Host
+			req.URL.Path = target.Path
+			req.Header.Set("Host", target.Host)
+		},
+	}
+	proxy.ServeHTTP(w, r)
+}
+
+func infoRefs(w http.ResponseWriter, r *http.Request) {
+	// We intercept the client request. We implement only the appropriate
+	// git-upload-pack service and enforce that clients are asking for this
+	// as well. The full specification of this service, as well as a formal
+	// grammar is available at:
+	//
+	//	https://gist.github.com/schacon/6092633#file-git-http-proto-txt-L224
```

Comment:
https://github.com/git/git/blob/master/Documentation/technical/http-protocol.txt seems to be the canonical version.